File: main/fraudulent-activity-notifications/fraudulent-activity-notifications.cs
#define DEBUG
//#define DEBUG_MEDIANS // checks computed medians with another (slow) method

using System;
using System.Diagnostics;
using System.Runtime.CompilerServices;
using static System.Diagnostics.Contracts.Contract;

[assembly: System.Reflection.AssemblyVersion("1.1.0.0")]

internal static class Solution {
    private const int MaxValue = 200;

    private static int[] ReadRecord()
        => Array.ConvertAll(Console.ReadLine().Split(), int.Parse);

    private static int[] ReadParameters(out int total_count,
                                        out int window_size)
    {
        var total_count_and_window_size = ReadRecord();
        total_count = total_count_and_window_size[0];
        window_size = total_count_and_window_size[1];
        Assert(0 < window_size && window_size <= total_count);

        var values = ReadRecord();
        Assert(values.Length == total_count);
        return values;
    }

    private static int[] GetInitialWindowCounts(this int[] values,
                                                int window_size)
    {
        var counts = new int[MaxValue + 1];
        for (var day = 0; day != window_size; ++day) ++counts[values[day]];
        return counts;
    }

    private static int CountTo(this int[] counts, int target, out int excess)
    {
        var value = 0;
        var running_count = 0;
        while ((running_count += counts[value]) < target) ++value;

        excess = running_count - target;
        return value;
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private static int CountToMiddle(this int[] counts,
                                     int window_size, out int excess)
        => counts.CountTo(window_size / 2 + 1, out excess);

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private static int ComputeDoubledMedian(this int[] counts,
            int window_size, int midval, int excess)
    {
        // The midpoint is exact when the number of elements is odd, or when it
        // is equal to the preceding value (i.e., with nonzero excess).
        if (excess != 0 || window_size % 2 != 0) return midval * 2;

        var value = midval - 1;
        while (counts[value] == 0) --value;
        return midval + value;
    }

    [Conditional("DEBUG_MEDIANS")]  
    private static void CheckDoubledMedian(this int[] values, int window_size,
                                           int day, int doubled_median)
    {
        const bool show_window_values = true;

        var window = new int[window_size];
        Array.Copy(values, day - window_size, window, 0, window_size);
        Array.Sort(window);

        var mid = window_size / 2;
        var result = (window_size % 2 == 0 ? window[mid] + window[mid - 1]
                                           : window[mid] * 2);

        Assert(doubled_median == result,
                $"{nameof(CheckDoubledMedian)} failed: "
                + (show_window_values ? $"({string.Join(" ", window)}) " : "")
                + $"{nameof(window_size)}={window_size}, {nameof(day)}={day}, "
                + $"fast={doubled_median}, slow={result}");
    }

    private static void Main()
    {
        int total_count, window_size;
        var values = ReadParameters(out total_count, out window_size);

        var counts = values.GetInitialWindowCounts(window_size);
        // TODO: computation of initial midval and excess will be moved here

        var notifications = 0;
        for (var day = window_size; day != total_count; ++day) {
            // get yesterday's midval and excess, so the median can be found
            // TODO: update/adjust these values instead of recomputing them
            int excess;
            var midval = counts.CountToMiddle(window_size, out excess);

            // check today's value against yesterday's median
            var doubled_median = counts.ComputeDoubledMedian(window_size,
                                                             midval, excess);
            values.CheckDoubledMedian(window_size, day, doubled_median);
            if (doubled_median <= values[day]) ++notifications;

            // update the counts (TODO: midval and excess will be updated here)
            ++counts[values[day]];
            --counts[values[day - window_size]];
        }

        Console.WriteLine(notifications);
    }
}

File: main/fraudulent-activity-notifications/fraudulent-activity-notifications.pl
#!/usr/bin/env perl

use strict;
use warnings;
use feature qw(say);
use experimental qw(lexical_subs);
use Carp qw(croak);
use List::Util qw(min);

use constant MAX_VALUE => 200;

sub read_record {
    return split /\s+/msx, <>;
}

my ($total_count, $window_size) = read_record;
my @values = read_record;
$total_count == scalar @values
    or croak q{total count doesn't match record length};
(0 < $window_size && $window_size <= $total_count)
    or croak q{unusable window size};

my $min_value = min @values;
0 <= $min_value or croak 'negative values are not allowed';

my @counts = (0) x (MAX_VALUE + 1);
for my $day (0 .. $window_size - 1) { ++$counts[$values[$day]] }

my $left_medial_count = int(($window_size + 1) / 2);
my $right_medial_count = int($window_size / 2) + 1;

(0 < $left_medial_count && $left_medial_count <= $right_medial_count
                        && $right_medial_count <= $left_medial_count + 1)
    or croak 'Internal error: computed medial counts are somehow wrong';

my sub compute_doubled_median {
    my $running_count = 0;
    my $value = $min_value - 1; # becomes the 1st value after initial increment

    my sub count_to {
        my ($target) = @_;

        while ($running_count < $target) {
            $running_count += $counts[++$value];
        }

        return $value;
    }

    my $left_medial_value = count_to $left_medial_count;
    my $right_medial_value = count_to $right_medial_count;
    return $left_medial_value + $right_medial_value;
}

my $notifications = 0;
for my $day ($window_size .. $total_count - 1) {
    # check today's value against the median _before_ sliding the window right
    if ($values[$day] >= compute_doubled_median) { ++$notifications }

    # update counts for the effect of sliding the window to the right
    ++$counts[$values[$day]];
    --$counts[$values[$day - $window_size]];
}

say $notifications;
